internal/kafka/internal/handlers/cluster.go

```diff
@@ -0,0 +1,98 @@
+package handlers
+
+import (
+	"net/http"
+
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/internal/kafka/internal/api/public"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/internal/kafka/internal/presenters"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/internal/kafka/internal/services"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/acl"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/api"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/errors"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/handlers"
+)
+
+type clusterHandler struct {
+	kasFleetshardOperatorAddon                           services.KasFleetshardOperatorAddon
+	clusterService                                       services.ClusterService
+	enterpriseClusterRegistrationAccessControlListConfig *acl.EnterpriseClusterRegistrationAccessControlListConfig
+}
+
+func NewClusterHandler(kasFleetshardOperatorAddon services.KasFleetshardOperatorAddon, clusterService services.ClusterService,
+	enterpriseClusterRegistrationAccessControlListConfig *acl.EnterpriseClusterRegistrationAccessControlListConfig) *clusterHandler {
+	return &clusterHandler{
+		kasFleetshardOperatorAddon: kasFleetshardOperatorAddon,
+		clusterService:             clusterService,
+		enterpriseClusterRegistrationAccessControlListConfig: enterpriseClusterRegistrationAccessControlListConfig,
+	}
+}
+
+func (h clusterHandler) RegisterEnterpriseCluster(w http.ResponseWriter, r *http.Request) {
+	var clusterPayload public.EnterpriseOsdClusterPayload
+
+	ctx := r.Context()
+
+	cfg := &handlers.HandlerConfig{
+		MarshalInto: &clusterPayload,
+		Validate: []handlers.Validate{
+			handlers.ValidateLength(&clusterPayload.ClusterId, "cluster id", ClusterIdLength, &ClusterIdLength),
+
+			handlers.ValidateExternalClusterId(&clusterPayload.ClusterExternalId, "external cluster id"),
+
+			handlers.ValidateClusterId(&clusterPayload.ClusterId, "cluster id"),
+
+			ValidateClusterIdIsUnique(&clusterPayload.ClusterId, h.clusterService),
+
+			handlers.ValidateDnsName(&clusterPayload.ClusterDnsName, "cluster dns name"),
+		},
+		Action: func() (interface{}, *errors.ServiceError) {
+
+			claims, claimsErr := getClaims(ctx)
+
+			if claimsErr != nil {
+				return nil, claimsErr
+			}
+
+			orgId, getOrgIdErr := claims.GetOrgId()
+
+			if getOrgIdErr != nil {
+				return nil, errors.GeneralError(getOrgIdErr.Error())
+			}
+
+			isOrgAccepted := h.enterpriseClusterRegistrationAccessControlListConfig.EnterpriseClusterRegistrationAccessControlList.IsOrganizationAccepted(orgId)
+
+			if !isOrgAccepted {
+				return nil, errors.GeneralError("organization '%s' is not authorized to register enterprise cluster", orgId)
+			}
```

Comment:
What do you think about extracting this logic quota onto a middleware that can be used only for this handler? 
Later on the middleware will deal with quota checks and the checks are performed before even landing to the handler

Reply:
do you mean something like `AccessControlListMiddleware`?

Reply:
Yes. Along those lines. But the new middleware will only be applied for this new endpoint.